### PR TITLE
raft: Avoid scanning raft log in become_leader

### PR DESCRIPTION
**Pull request by csmoe**
should resolve issue #11 
modifications based on pull request https://github.com/coreos/etcd/pull/9073

### CODE REVIEW

**Comment by siddontang** [comment]
Great work @csmoe 

PTAL @Hoverbear @BusyJay

File: src/raft.rs [hunk 1]
```diff
@@ -780,13 +785,13 @@ impl<T: Storage> Raft<T> {
...
-
-        if nconf == 1 {
-            self.pending_conf = true;
+        // Conservatively set the pending_conf_index to the last index in the
+        // log. There may or may not be a pending config change, but it's
+        // safe to delay any future proposals until we commit all our
+        // pending log entries, and scanning the entire tail of the log
+        // could be expensive.
+        if ents.len() > 0 {
+            self.pending_conf_index = ents[0].get_index() + ents.len() as u64 - 1;
```
Comment (BusyJay) [inline src/raft.rs]: You can use `ents.last().map_or(0, |e| e.get_index())`.

File: src/raft.rs [hunk 2]
```diff
@@ -1819,7 +1826,8 @@ impl<T: Storage> Raft<T> {
     }
 
     pub fn should_bcast_commit(&self) -> bool {
-        !self.skip_bcast_commit || self.pending_conf
+        let last_index = self.raft_log.last_index();
+        !self.skip_bcast_commit || (self.pending_conf_index == last_index)
```
Comment (BusyJay) [inline src/raft.rs]: I think it should be `self.pending_conf_index > self.raft_log.applied`.

File: src/raft.rs [hunk 3]
```diff
@@ -780,13 +785,13 @@ impl<T: Storage> Raft<T> {
...
-        }
-
-        if nconf == 1 {
-            self.pending_conf = true;
+        // Conservatively set the pending_conf_index to the last index in the
+        // log. There may or may not be a pending config change, but it's
+        // safe to delay any future proposals until we commit all our
+        // pending log entries, and scanning the entire tail of the log
+        // could be expensive.
+        if ents.len() > 0 {
```
Comment (BusyJay) [inline src/raft.rs]: The check is unnecessary anymore.

File: src/raft.rs [hunk 4]
```diff
@@ -1819,7 +1826,7 @@ impl<T: Storage> Raft<T> {
     }
 
     pub fn should_bcast_commit(&self) -> bool {
-        !self.skip_bcast_commit || self.pending_conf
+        !self.skip_bcast_commit || (self.pending_conf_index > self.raft_log.applied)
```
Comment (BusyJay) [inline src/raft.rs]: Any tests to cover this?

**Comment by siddontang** [comment]
any update @csmoe

**Comment by csmoe** [comment]
@siddontang  sorry for the delay. I'm trying to figure out the tests for `self.pending_conf_index > self.raft_log.applied` in `pub fn should_bcast_commit(&self) -> bool`

File: tests/cases/test_raw_node.rs [hunk 5]
```diff
@@ -429,6 +429,11 @@ fn test_skip_bcast_commit() {
     // elect r1 as leader
     nt.send(vec![new_message(1, 1, MessageType::MsgHup, 0)]);
 
+    // Skip bcast commit when pending_conf_index <= applied
```
Comment (siddontang) [inline tests/cases/test_raw_node.rs]: can we check the result of `should_bcast_commit` here?

**Comment by siddontang** [comment]
LGTM

PTAL @BusyJay @hicqu

**Comment by siddontang** [comment]
PTAL @BusyJay @hicqu

**Pull request merged**